rust/rust.cc

```diff
@@ -0,0 +1,323 @@
+    float npnr_context_get_pip_delay(const Context *const ctx, uint64_t pip) { return ctx->getDelayNS(ctx->getPipDelay(unwrap_pip(pip)).maxDelay()); }
+    float npnr_context_get_wire_delay(const Context *const ctx, uint64_t wire) { return ctx->getDelayNS(ctx->getWireDelay(unwrap_wire(wire)).maxDelay()); }
+    float npnr_context_delay_epsilon(const Context *const ctx) { return ctx->getDelayNS(ctx->getDelayEpsilon()); }
+    Loc npnr_context_get_pip_location(const Context *const ctx, uint64_t pip) { return ctx->getPipLocation(unwrap_pip(pip)); }
+    bool npnr_context_check_pip_avail_for_net(const Context *const ctx, uint64_t pip, NetInfo *net) { return ctx->checkPipAvailForNet(unwrap_pip(pip), net); }
+
+    // This method's in C++ temporarily, while I figure out some better way of getting a pip iterator.
+    Loc npnr_context_get_pip_direction(const Context *const ctx, uint64_t _pip) {
+        auto pip = unwrap_pip(_pip);
+        auto src_loc = Loc{};
+        auto dst_loc = Loc{};
+
+        auto uh_pips = 0;
+        for (auto uh : ctx->getPipsUphill(ctx->getPipSrcWire(pip))) {
+            auto loc = ctx->getPipLocation(uh);
+            src_loc.x += loc.x;
+            src_loc.y += loc.y;
+            uh_pips++;
+        }
+        if (uh_pips > 1) {
+            src_loc.x /= uh_pips;
+            src_loc.y /= uh_pips;
+        }
+
+        auto dh_pips = 0;
+        for (auto dh : ctx->getPipsDownhill(ctx->getPipDstWire(pip))) {
+            auto loc = ctx->getPipLocation(dh);
+            dst_loc.x += loc.x;
+            dst_loc.y += loc.y;
+            dh_pips++;
+        }
+        if (dh_pips > 1) {
+            dst_loc.x /= dh_pips;
+            dst_loc.y /= dh_pips;
+        }
+
+        dst_loc.x -= src_loc.x;
+        dst_loc.y -= src_loc.y;
+        return dst_loc;
+    }
+
+    uint64_t npnr_context_get_pips_leak(const Context *const ctx, PipId **pips) {
+        auto pip_vec = std::vector<PipId>{};
+        for (auto pip : ctx->getPips()) {
+            pip_vec.push_back(pip);
+        }
+        pip_vec.shrink_to_fit();
+        auto size = pip_vec.size();
+        *pips = pip_vec.data();
+        auto dummy = std::vector<PipId>{};
+        // Yes, by memcpying over `pip_vec` we leak memory.
+        std::memcpy(&pip_vec, &dummy, sizeof(dummy));
+        return size;
+    }
+
+    uint64_t npnr_context_get_wires_leak(const Context *const ctx, WireId **wires) {
+        auto wire_vec = std::vector<WireId>{};
+        for (auto wire : ctx->getWires()) {
+            wire_vec.push_back(wire);
+        }
+        wire_vec.shrink_to_fit();
+        auto size = wire_vec.size();
+        *wires = wire_vec.data();
+        auto dummy = std::vector<WireId>{};
+        // Yes, by memcpying over `wire_vec` we leak memory.
+        std::memcpy(&wire_vec, &dummy, sizeof(dummy));
```

Comment:
Could you avoid the memcpy situation here by doing something like `std::vector<WireId> *wire_vec = new std::vector<WireId>{};` `...` and then just never deleting it? That way the vector will also just stay around forever.

Reply:
Another option: iterate over wires once to get the number of wires; allocate an array of WireId of that size; and then iterate again to populate it? This could well be faster than a vector dynamically resizing itself several times; and also means ownership of the allocated memory can be cleanly passed to Rust that could safely delete it if it so desired (and such a thing is possible?)

Final thought here: should this be a list of `uint64_t` wrapped WireIds rather than just a list of WireIds?